### ISSUE TITLE
test error on windows

### ISSUE THREAD

**Issue by guillermooo** [dart-lang/linter#128]
```
ERROR: rule | pub | package_names
  Test failed: Caught Unsupported operation: Illegal character in path
  dart:core/uri.dart 769                                 Uri._checkWindowsPathReservedCharacters
  dart:core/uri.dart 1833                                Uri._toWindowsFilePath
  dart:core/uri.dart 1807                                Uri.toFilePath
  package:linter/src/analysis.dart 29:48                 createSource
  package:linter/src/linter.dart 220:21                  LintRule.reportPubLint
  package:linter/src/rules/pub/package_names.dart 46:12  Visitor.visitPackageName
  package:linter/src/pub.dart 328:15                     _Pubspec.accept
  package:linter/src/linter.dart 346:18                  SourceLinter.lintPubspecSource
  package:linter/src/linter.dart 365:7                   SourceLinter._lintPubspecFile
  package:linter/src/linter.dart 321:28                  SourceLinter.lintFiles.<fn>
  dart:core/iterable.dart 217                            Iterable.forEach
  package:linter/src/linter.dart 319:42                  SourceLinter.lintFiles
  test\linter_test.dart 508:48                           testRule.<fn>
  package:unittest/src/internal_test_case.dart 120:37    InternalTestCase.run.<fn>
  dart:async/zone.dart 911                               _rootRunUnary
  dart:async/zone.dart 811                               _CustomZone.runUnary
  dart:async/future_impl.dart 494                        _Future._propagateToListeners.handleValueCallback
  dart:async/future_impl.dart 577                        _Future._propagateToListeners
  dart:async/future_impl.dart 368                        _Future._completeWithValue
  dart:async/future_impl.dart 422                        _Future._asyncComplete.<fn>
  dart:async/zone.dart 904                               _rootRun
  dart:async/zone.dart 803                               _CustomZone.run
  dart:async/zone.dart 709                               _CustomZone.runGuarded
  dart:async/zone.dart 734                               _CustomZone.bindCallback.<fn>
  dart:async/schedule_microtask.dart 43                  _microtaskLoop
  dart:async/schedule_microtask.dart 52                  _microtaskLoopEntry
  dart:isolate-patch/timer_impl.dart 394                 _Timer._runTimers
  dart:isolate-patch/timer_impl.dart 414                 _Timer._handleMessage
  dart:isolate-patch/isolate_patch.dart 148              _RawReceivePortImpl._handleMessage
```

It seems this fails ultimately because of this:

```dart
void main() {
  var uFile = new Uri.file(r"test\rules\pub\package_names\_pubspec.yaml");
  var uParse = Uri.parse(r"test\rules\pub\package_names\_pubspec.yaml");
  print(uFile.toFilePath());
  print(uParse.toFilePath()); // exception on Windows
}
```

I stopped investigating here. Could this be a bug in `Uri` ([Uri code](https://github.com/dart-lang/sdk/blob/de300cfa0e24aa33559a8625350ae86ca6d1b389/sdk/lib/core/uri.dart#L769))?

**Comment by guillermooo** [comment]
https://github.com/dart-lang/linter/issues/113#issuecomment-133087858

**Comment by guillermooo** [comment]
I think at least two changes are required to fix this error:

One is SourceSpan/file.dart

```dart
  /// Creates a new source file from a list of decoded characters.
  ///
  /// [url] may be either a [String], a [Uri], or `null`.
  SourceFile.decoded(Iterable<int> decodedChars, {url})
      : url = url is String ? Uri.parse(url.replaceAll(r'\', r'/')) : url, // <== avoid \ here
        _decodedChars = new Uint32List.fromList(decodedChars.toList()) {
    for (var i = 0; i < _decodedChars.length; i++) {
      var c = _decodedChars[i];
      if (c == _CR) {
        // Return not followed by newline is treated as a newline
        var j = i + 1;
        if (j >= _decodedChars.length || _decodedChars[j] != _LF) c = _LF;
      }
      if (c == _LF) _lineStarts.add(i + 1);
    }
  }
```

I'm suggesting this change to a different package because I believe this situation will always be a potential problem on Windows, and the replacement doesn't change the semantics of the path of Windows AFAIK.

Second change in src/linter.dart in this repo:

```dart
  @override
  Iterable<AnalysisErrorInfo> lintPubspecSource(
      {String contents, String sourceUrl}) {
    var results = <AnalysisErrorInfo>[];

    var spec = new Pubspec.parse(contents, sourceUrl: sourceUrl);

    for (Linter lint in options.enabledLints) {
      if (lint is LintRule) {
        LintRule rule = lint;
        var visitor = rule.getPubspecVisitor();
        if (visitor != null) {
          // Analyzer sets reporters; if this file is not being analyzed,
          // we need to set one ourselves.  (Needless to say, when pubspec
          // processing gets pushed down, this hack can go away.)
          if (rule.reporter == null && sourceUrl != null) {
            sourceUrl = sourceUrl.replaceAll(r'\', r'/'); // avoid \ when Uri.parse'ing
            var source = createSource(Uri.parse(sourceUrl));
            rule.reporter = new ErrorReporter(this, source);
          }
          try {
            spec.accept(visitor);
          } on Exception catch (e) {
            reporter.exception(new LinterException(e.toString()));
          }
          if (rule._locationInfo != null && !rule._locationInfo.isEmpty) {
            results.addAll(rule._locationInfo);
            rule._locationInfo.clear();
          }
        }
      }
    }

    return results;
  }
```

I don't know if this is treating the symptons only, but at least it's some sort of progress to fix the error.

```
unittest-suite-wait-for-done
test/rules/pub/package_names/_pubspec.yaml
PASS: rule | pub | package_names
```

**Comment by guillermooo** [comment]
Some more digging — `Uri.parse` does not seem to support UNC paths (not that it should, I guess), so that's one more reason to replace all `\` in this case? Also, if `new Uri.file` is used instead to instantiate the URI *and* we perform the replacement as well, the `Uri` members that return a path will return it with backslashes in any case.

**Comment by pq** [comment]
Thanks for the digging @guillermooo. :+1: 

Since this touches on the `span` bits, I'm curious how @nex3 weighs in...

**Comment by nex3** [comment]
First:

> ```dart
> void main() {
>   var uFile = new Uri.file(r"test\rules\pub\package_names\_pubspec.yaml");
>   var uParse = Uri.parse(r"test\rules\pub\package_names\_pubspec.yaml");
>   print(uFile.toFilePath());
>   print(uParse.toFilePath()); // exception on Windows
> }
> ```

This is working correctly. `Uri.parse` parses a *URI*, which is different than a *path*. `\` is not a valid separator for URIs, so the parsing code just treats it as a normal character in a very long path name. However, `\` *is* a separator on Windows and thus not a valid filename character, so there's no way to convert this URI to a Windows path and `Uri.toFilePath` throws.

You should essentially never be passing `\` to code expecting a URI. If you have a path, you should use `new Uri.file` or [`p.toUri`](http://www.dartdocs.org/documentation/path/1.3.6/index.html#path/path@id_toUri) before you pass it to an API that wants a URI. Otherwise you'll see issues like this.

Digging through the stack trace, it looks like the root of this issue is [the `_lintPubspecFile` function](https://github.com/dart-lang/linter/blob/master/lib/src/linter.dart#L364-L366), which passes a path to a function that expects a URI.

**Comment by guillermooo** [comment]
Thanks for the additional info. I did some tests agains the Uri class in the .Net framework and I seem to remember that it does things differently than the Uri class in Dart. Perhaps on Windows the .Net implementation of Uri it could be taken as reference? I think the main difference is that the .Net Uri has a `ToLocalPath()` method that is more lenient in accepting `\` in Uris. This is probably technically wrong, but nevertheless... Anyway, take this with a grain of salt! I'm speaking from memory.

**Comment by guillermooo** [comment]
Hm. I misrimembered. It seems that .Net Uri won't not parse at all the path above, not even escaping the backslashes in ToString().

**Comment by pq** [comment]
Thanks for chiming in @nex3!  I think [path.toUri](http://www.dartdocs.org/documentation/path/1.3.6/index.html#path/path@id_toUri) is exactly what we want.

**Comment by pq** [comment]
Fixed w/ a17423da72fc2d945e08d36ca7e6b74ca00231b1.
